src/Report/EnabledMembers.php

```diff
@@ -0,0 +1,167 @@
+use SilverStripe\Forms\NumericField;
+use SilverStripe\Forms\TextField;
+use SilverStripe\MFA\Method\MethodInterface;
+use SilverStripe\MFA\Service\MethodRegistry;
+use SilverStripe\ORM\ArrayList;
+use SilverStripe\ORM\SS_List;
+use SilverStripe\Reports\Report;
+use SilverStripe\Security\Member;
+
+if (!class_exists(Report::class)) {
+    return;
+}
+
+class EnabledMembers extends Report
+{
+    protected $title = 'Members with MFA enabled on their account';
+
+    protected $description = 'Identifies all members on this site, with columns added to indicate the status each has'
+        . ' in regards to multi factor authentication method registration. E.g. how many methods are set up by that'
+        . ' member, which method types, and who has opted not to set up MFA (if this option is enabled for the site)';
+
+    protected $dataClass = Member::class;
+
+    public function title()
+    {
+        return _t(__CLASS__ . '.TITLE', parent::title());
+    }
+
+    public function description()
+    {
+        return _t(__CLASS__ . '.DESCRIPTION', parent::description());
+    }
+
+    /**
+     * @param array $params
+     */
+    public function sourceRecords($params)
+    {
+        $methods = $this->getMethodClassToTitleMapping();
+
+        // Having a WHERE limits the results, which means the the COUNT will be off (GROUP BY applied after WHERE)
+        // So clip the count filter and apply it after the database query, only if Methods filter is also set
+        // this way we can delegate to the database where possible, which should be faster.
+        if (isset($params['Count'])) {
+            $desiredCount = (int)$params['Count'];
+            if (isset($params['Methods'])) {
+                $countFilter = $desiredCount;
+                unset($params['Count']);
+            } elseif ($desiredCount) {
+                $params['Count'] = $desiredCount++;
+            }
+        }
+
+        $members = $this->applyParams(Member::get(), $params)->toArray();
+        array_walk($members, function ($member) use ($methods) {
+            $defaultMethod = $member->getDefaultRegisteredMethod();
+            if ($defaultMethod) {
+                $defaultMethodClassName = $defaultMethod ? $defaultMethod->MethodClassName : null;
+                $backupClass = get_class(Injector::inst()->get(MethodRegistry::class)->getBackupMethod());
+                $registeredMethods = $member
+                    ->RegisteredMFAMethods()
+                    ->exclude('MethodClassName', $backupClass)
+                    ->column('MethodClassName');
+                $registeredMethodNames = array_map(function ($methodClass) use ($methods) {
+                    return $methods[$methodClass];
+                }, $registeredMethods);
+                $registeredMethodList = implode(', ', $registeredMethodNames);
+                $methodCount = count($registeredMethods);
+                $member->methodCount = $methodCount ?: '';
```

Comment:
The direct modification/random property creation here feels dirty. Can we create `ArrayData` objects for the report to render instead?

Reply:
Yes.

Reply:
The logic here seems a bit odd to me - you assign `$methodCount`'s value (an int) or an empty string - can you not just do `= (string) count($registeredMethods)`?

Reply:
That will return `"0"`, which is not what I want here (I want an empty string).